File: AndroidClipSync/app/src/main/java/com/mtechdwork/clipsync/BroadcastHandler.java
package com.mtechdwork.clipsync;

import android.content.Context;
import android.net.wifi.WifiInfo;
import android.net.wifi.WifiManager;
import android.util.Log;

import java.net.DatagramPacket;
import java.net.DatagramSocket;
import java.net.InetAddress;
import java.util.Objects;

public class BroadcastHandler extends Thread {

    private static final int PORT = 7070;
    private boolean running = true;
    private Context context;
    private SettingManager settingManager;

    private DatagramSocket socket;

    BroadcastHandler(Context context) {
        this.context = context;
        settingManager = new SettingManager(context);
    }

    private void log(String message, int type) { // Debug method
        // Type: 0 - Info, 1 - Warning, 2 - Error
        boolean debug = true;
        if (!debug) return;
        String className = "[Broadcast Handler]";
        switch (type) {
            case 0:
                Log.i(className, message);
                break;

            case 1:
                Log.w(className, message);
                break;

            case 2:
                Log.e(className, message);
        }
    }

    public void stopListening() {
        running = false;
        if (socket != null && !socket.isClosed()) socket.close();
    }

    private String getIPAddress() {
        WifiManager wifiManager = (WifiManager) context.getApplicationContext().getSystemService(Context.WIFI_SERVICE);
        if (wifiManager != null) {
            WifiInfo wifiInfo = wifiManager.getConnectionInfo();
            int ipAddress = wifiInfo.getIpAddress();

            // Chuyển đổi từ int sang địa chỉ IP
            return (ipAddress & 0xFF) + "." +
                    ((ipAddress >> 8) & 0xFF) + "." +
                    ((ipAddress >> 16) & 0xFF) + "." +
                    ((ipAddress >> 24) & 0xFF);
        }
        return null;
    }

    private boolean checkSenderMatch(String receivedMessage) {
        String broadcastContent = "CS_BC_" + settingManager.getUsername();
        return receivedMessage.equals(broadcastContent);
    }

    @Override
    public void run() {
        try {
            socket = new DatagramSocket(PORT);
            byte[] buffer = new byte[1024];

            while (running) {
                DatagramPacket packet = new DatagramPacket(buffer, buffer.length);
                socket.receive(packet);

                String receivedMessage = new String(packet.getData(), 0, packet.getLength());
                InetAddress senderAddress = packet.getAddress();

                if (checkSenderMatch(receivedMessage) && !Objects.equals(senderAddress.getHostAddress(), getIPAddress())) {
                    log("Received: " + receivedMessage + " from " + senderAddress.getHostAddress(), 0);
                    Communication communication = new Communication(context);
                    communication.sendSyncRequest(senderAddress);
                }
            }

            if (socket != null) socket.close();
        } catch (Exception e) {
            log("EXCEPTION", 2);
            e.printStackTrace();
            stopListening();
        }
        log("Thread stopped!", 0);
    }

    @Override
    public void interrupt() {
        super.interrupt();
        log("Thread interrupt!", 0);
        stopListening();
    }
}

File: WindowsClipSync/ClipSync/Classes/BroadcastHandler.vb
Imports System.Net
Imports System.Net.Sockets
Imports System.Text
Imports System.Threading

Public Class BroadcastHandler
	Private broadcastListening As Thread
	Private listener As UdpClient
	Private running = False
	Private UDP_PORT = 7070

	Public Sub startListening()
		running = True
		listener = New UdpClient(New IPEndPoint(LocalIP.getIPAddress(), UDP_PORT))
		broadcastListening = New Thread(AddressOf listening)
		broadcastListening.IsBackground = True
		broadcastListening.Start()
	End Sub

	Public Sub stopListening()
		running = False
		If listener IsNot Nothing Then listener.Close()
	End Sub

	Private Function checkSenderMatch(receivedMessage As String) As Boolean
		Dim broadcastContent = "CS_BC_" + SettingManager.getUsername()
		Return receivedMessage = broadcastContent
	End Function

	Private Sub listening()
		ClipSyncDebug.log("[BroadcastHandler] Start new thread")

		Dim remoteEP As New IPEndPoint(IPAddress.Any, 0)
		While running
			Try
				Dim data As Byte() = listener.Receive(remoteEP)
				Dim message As String = Encoding.UTF8.GetString(data)

				If checkSenderMatch(message) AndAlso LocalIP.getIPAddress().ToString <> remoteEP.Address.ToString Then
					ClipSyncDebug.log("Received from " & remoteEP.Address.ToString() & ": " & message)
					Communication.sendSyncRequest(remoteEP.Address)
				End If
			Catch ex As Exception
				ClipSyncDebug.log("[BroadcastHandler] UDP receive error: " & ex.Message)
			End Try
		End While
		ClipSyncDebug.log("[BroadcastHandler] Stop thread")
	End Sub
End Class
